treeherder/etl/taskcluster_pulse/handler.py

```diff
@@ -0,0 +1,302 @@
+root_url = "https://taskcluster.net"
+
+
+class PulseHandlerError(Exception):
+    """Base error"""
+    pass
+
+
+def stateFromRun(jobRun):
+    return "completed" if jobRun["state"] in ("exception", "failed") else jobRun["state"]
+
+
+def resultFromRun(jobRun):
+    if jobRun["state"] == "completed":
+        return "success"
+    elif jobRun["state"] == "failed":
+        return "fail"
+    elif jobRun["state"] == "exception":
+        if jobRun["reasonResolved"] == "canceled":
+            return "canceled"
+
+        if jobRun["reasonResolved"] == "superseded":
+            return "superseded"
+
+        return "exception"
+    else:
+        return "unknown"
+
+
+# Creates a log entry for Treeherder to retrieve and parse.  This log is
+# displayed on the Treeherder Log Viewer once parsed.
+def createLogReference(taskId, runId):
+    logUrl = taskcluster_urls.api(
+        root_url,
+        "queue",
+        "v1",
+        "task/{taskId}/runs/{runId}/artifacts/public/logs/live_backing.log"
+    ).format(taskId=taskId, runId=runId)
+    return {
+        # XXX: This is a magical name see 1147958 which enables the log viewer.
+        "name": "builds-4h",
+        "url": logUrl,
+    }
+
+
+# Filters the task routes for the treeherder specific route.  Once found,
+# the route is parsed into distinct parts used for constructing the
+# Treeherder job message.
+# XXX: This function is a perfect candidate for some refactoring
+def parseRouteInfo(prefix, taskId, routes, task):
+    matchingRoutes = list(filter(lambda route: route.split(".")[0] == "tc-treeherder", routes))
+
+    if len(matchingRoutes) != 1:
+        raise PulseHandlerError(
+            "Could not determine Treeherder route. Either there is no route, " +
+            "or more than one matching route exists." +
+            "Task ID: {taskId} Routes: {routes}".format(taskId=taskId, routes=routes)
+        )
+
+    parsedRoute = parseRoute(matchingRoutes[0])
+    # During a transition period, some tasks might contain a revision within
+    # the task definition that should override the revision in the routing key.
+    revision = task.get("extra", {}).get("treeherder", {}).get("revision")
+
+    if revision:
+        parsedRoute["revision"] = revision
+
+    return parsedRoute
+
+
+def validateMessage(taskMetadata):
+    treeherderMetadata = taskMetadata.get("extra", {}).get("treeherder")
+    if not treeherderMetadata:
+        logger.debug("Task metadata is missing Treeherder job configuration.")
+        return False
+    try:
+        jsonschema.validate(treeherderMetadata, get_json_schema("task-treeherder-config.yml"))
+    except (jsonschema.ValidationError, jsonschema.SchemaError) as e:
+        logger.error("JSON Schema validation error during Taskcluser message ingestion: %s", e)
+        return False
+    return True
+
+
+def fetchMetadata(taskId):
+    url = taskcluster_urls.api(root_url, "queue", "v1", "task/{taskId}".format(taskId=taskId))
+    return fetch_json(url)
```

Comment:
This is a bit of a confusing function name -- it's fetching the task definition.  That happens to contain a `metadata` field but it's probably not what you want (`name`, `description`, etc.).  I'd call this function `fetchTask`.

While this approach of constructing a URL manually and fetching it is fine, you could also use a `taskcluster.Queue` object's `task` method to do the same.  The latter comes with some retry logic that `fetch_json` might not have.